doc/man3/OSSL_HPKE_CTX_new.pod

```diff
@@ -0,0 +1,328 @@
+defined in hpke.h. The suite strings recognised are comma-separated KEM, KDF
+and AEAD names in that order, e.g. "x25519,hkdf-sha256,aes128gcm". This can
+be used by e.g., command line tools that accept string form names for 
+HPKE codepoints.
+
+=head2 Key Generation
+
+Any OpenSSL-supported method producing an EVP_PKEY for generating keys that
+match the IANA-registered codepoints for HPKE may be used. We also provide
+a specific API that is aware of how those codepoints map to OpenSSL algorithm
+implementations.
+
+OSSL_HPKE_keygen() generates an ECDH private value for use with the
+given HPKE suite. This returns the private key as an
+EVP_PKEY*. Only the kem_id field of the suite input is currently used. If the
+initial key material (IKM) inputs (ikmlen/ikm) are (0/NULL) then a key pair
+will be randomly generated. If those input are nonzero/non-NULL then the 
+deterministic key generation scheme from RFC9180 will be used.
+
+=head2 OSSL_HPKE_CTX manipulating APIs
+
+Callers can construct an OSSL_HPKE_CTX object using OSSL_HPKE_CTX_new() 
+that removes the need to provide repeated input parameters such as the
+libctx, propq, HPKE mode and suite. The OSSL_HPKE_CTX_free()
+function releases that memory.
+
+Three additional attributes may be set for such context objects: a 
+pre-shared key (PSK), a sender authentication key and to control (via set/get)
+the sequence number for the next seal/open operation.
+
+Sender key generation can also be based on the deterministic key generation
+scheme specified in RFC 9180 via the OSSL_HPKE_CTX_set1_ikme() function.
+
+If the "EXPORT-ONLY" mode from RFC 9180 is desired, then OSSL_HPKE_export()
+can be used after a call to OSSL_HPKE_sender_export_encap() or
+OSSL_HPKE_recipient_export_decap() if the relevant export-only AEAD has been
+chosen (OSSL_HPKE_AEAD_ID_EXPORTONLY).
+
+=head1 RETURN VALUES
+
+All functions return 1 for success. Functions may return a negative 
+number, zero or a positive number in the event of an error.
+
+=head1 EXAMPLES
+
+This example demonstrates a minimal round-trip using HPKE.
+
+    #include <openssl/hpke.h>
+    /* we'll do a round-trip, generating a key, encrypting and decrypting */
+    int hpke_mode=OSSL_HPKE_MODE_BASE;
+    OSSL_HPKE_SUITE hpke_suite = OSSL_HPKE_SUITE_DEFAULT;
+    OSSL_HPKE_CTX *ctx = NULL, *rctx = NULL;
+    size_t publen=OSSL_HPKE_MAXSIZE; unsigned char pub[OSSL_HPKE_MAXSIZE];
+    EVP_PKEY *priv = NULL;
+    size_t enclen=OSSL_HPKE_MAXSIZE; unsigned char enc[OSSL_HPKE_MAXSIZE];
+    size_t plainlen=OSSL_HPKE_MAXSIZE; unsigned char plain[OSSL_HPKE_MAXSIZE];
+    size_t cipherlen=OSSL_HPKE_MAXSIZE; unsigned char cipher[OSSL_HPKE_MAXSIZE];
+    size_t ptlen=OSSL_HPKE_MAXSIZE; unsigned char pt[OSSL_HPKE_MAXSIZE];
+
+    memset(pt,0,OSSL_HPKE_MAXSIZE);
+    strcpy((char*)pt,"a message not in a bottle");
+    ptlen=strlen((char*)pt);
+
+    /* generate receiver's key pair */
+    OSSL_HPKE_keygen(NULL, NULL, hpke_mode, hpke_suite,
```

Comment:
should this do a little bit of error checking.. Other examples do goto error..

Reply:
sure, included a fully working program for now, happy to elide whatever people want as we go